### DISCUSSION TITLE
Linux netstat for mac

### DISCUSSION THREAD

**Discussion by EricoCartmanez** [Homebrew/discussions#860] [Everyday usage] [answered]
Hi,

I think it would be great if we could somehow install **netstat** used in linux distributions via brew as mac's netstat it's almost useless. 

Thank you.

**Comment by Bo98** [comment]
`netstat` on Linux is now deprecated/obsolete, replaced by `ss` in iproute2.

I don't know much about the old net-tools, but iproute2 definitely relies on the Linux kernel and unfortunately can't be trivially built on macOS.

**Reply by EricoCartmanez** [reply]
**SS** then :)
You can't install SS via brew either. 

P.S.: Netstat is supposed to be deprecated since years but everybody's using it.

**Comment by SMillerDev** [comment]
https://docs.brew.sh/Adding-Software-to-Homebrew would be the best place to start.

**Reply by EricoCartmanez** [reply]
A lot to read there... Anyway I'm comparing existing formulas to get an idea. 

How can I know what packages the program depends on?

For example from wget formula:
```
   head do
   url "https://git.savannah.gnu.org/git/wget.git"
 
     depends_on "autoconf" => :build
     depends_on "automake" => :build
     depends_on "xz" => :build
     depends_on "gettext"
   end
 
 depends_on "pkg-config" => :build
 depends_on "libidn2"
 depends_on "openssl@1.1"
```

On linux with "apt show wget" you see:
```
Depends: libc6 (>= 2.17), libidn2-0 (>= 0.6), libpcre2-8-0 (>= 10.32), libpsl5 (>= 0.16.0), libssl1.1 (>= 1.1.0), libuuid1 (>= 2.16), zlib1g (>= 1:1.1.4)
```

**Reply by SMillerDev** [reply]
You'll have to read the upstream install documentation.

**Reply by EricoCartmanez** [reply]
Maybe you guys should make a video...

**Reply by SMillerDev** [reply]
A video on how to read the documentation from the project you're trying to build?

**Reply by EricoCartmanez** [reply]
A video of how to import a linux package to brew. As part of the documentation.

**Reply by SMillerDev** [reply]
There is no one way to add software to brew, every software has it's own dependencies and requirements. The only real information you can get is usually provided by the upstream project.

**Reply by EricoCartmanez** [reply]
Trying to compile it on my machine. I manually created some missing files as suggested  [here](https://sourceforge.net/p/e2fsprogs/bugs/311/) but..
I don't think it will ever work :/

```
my@macbook  ~/Desktop/iproute2-5.10.0  make

lib
    CC       libgenl.o
In file included from libgenl.c:10:
In file included from ../include/uapi/linux/genetlink.h:5:
../include/uapi/linux/types.h:5:10: error: 'asm/types.h' file not found with <angled> include; use "quotes" instead
#include <asm/types.h>
         ^~~~~~~~~~~~~
         "asm/types.h"
In file included from libgenl.c:10:
In file included from ../include/uapi/linux/genetlink.h:5:
In file included from ../include/uapi/linux/types.h:9:
../include/uapi/linux/posix_types.h:36:10: error: 'asm/posix_types.h' file not found with <angled> include; use "quotes" instead
#include <asm/posix_types.h>
         ^~~~~~~~~~~~~~~~~~~
         "asm/posix_types.h"
In file included from libgenl.c:10:
In file included from ../include/uapi/linux/genetlink.h:5:
../include/uapi/linux/types.h:24:9: error: unknown type name '__u16'
typedef __u16 __bitwise __le16;
        ^
../include/uapi/linux/types.h:25:9: error: unknown type name '__u16'
typedef __u16 __bitwise __be16;
        ^
../include/uapi/linux/types.h:26:9: error: unknown type name '__u32'
typedef __u32 __bitwise __le32;
        ^
../include/uapi/linux/types.h:27:9: error: unknown type name '__u32'
typedef __u32 __bitwise __be32;
        ^
../include/uapi/linux/types.h:28:9: error: unknown type name '__u64'
typedef __u64 __bitwise __le64;
        ^
../include/uapi/linux/types.h:29:9: error: unknown type name '__u64'
typedef __u64 __bitwise __be64;
        ^
../include/uapi/linux/types.h:31:9: error: unknown type name '__u16'
typedef __u16 __bitwise __sum16;
        ^
../include/uapi/linux/types.h:32:9: error: unknown type name '__u32'
typedef __u32 __bitwise __wsum;
        ^
In file included from libgenl.c:10:
In file included from ../include/uapi/linux/genetlink.h:6:
In file included from ../include/uapi/linux/netlink.h:5:
In file included from ../include/uapi/linux/kernel.h:5:
../include/uapi/linux/sysinfo.h:9:2: error: unknown type name '__kernel_long_t'
        __kernel_long_t uptime;         /* Seconds since boot */
        ^
../include/uapi/linux/sysinfo.h:10:2: error: unknown type name '__kernel_ulong_t'
        __kernel_ulong_t loads[3];      /* 1, 5, and 15 minute load averages */
        ^
../include/uapi/linux/sysinfo.h:11:2: error: unknown type name '__kernel_ulong_t'
        __kernel_ulong_t totalram;      /* Total usable main memory size */
        ^
../include/uapi/linux/sysinfo.h:12:2: error: unknown type name '__kernel_ulong_t'
        __kernel_ulong_t freeram;       /* Available memory size */
        ^
../include/uapi/linux/sysinfo.h:13:2: error: unknown type name '__kernel_ulong_t'
        __kernel_ulong_t sharedram;     /* Amount of shared memory */
        ^
../include/uapi/linux/sysinfo.h:14:2: error: unknown type name '__kernel_ulong_t'
        __kernel_ulong_t bufferram;     /* Memory used by buffers */
        ^
../include/uapi/linux/sysinfo.h:15:2: error: unknown type name '__kernel_ulong_t'
        __kernel_ulong_t totalswap;     /* Total swap space size */
        ^
../include/uapi/linux/sysinfo.h:16:2: error: unknown type name '__kernel_ulong_t'
        __kernel_ulong_t freeswap;      /* swap space still available */
        ^
../include/uapi/linux/sysinfo.h:17:2: error: unknown type name '__u16'
        __u16 procs;                    /* Number of current processes */
        ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
20 errors generated.
```

**Reply by danielnachun** [reply]
Those errors are happening because `iproute2` needs the Linux API, which means it needs the Linux kernel.  So you are correct that it will probably never work on macOS.  It's unfortunate that there doesn't seem to be an equivalent tool on macOS.  

The only solution I can think of is to see if you can set up `iproute2` in a Linux container and see if you can forward the network connections to the VM.  I am not a network expert, even less when it comes to VMs, so I don't know if the virtualization would interfere with whatever you need to do with `iproute2`.  But that might be more convenient than having to dual boot Linux or run it on a different machine.

**Reply by gwerbin** [reply]
There is an alternative implementation of `sockstat` that ships with FreeBSD, but it might depend on parts of the BSD kernel which are not shared by XNU (the MacOS kernel).

**Reply by mezantrop** [reply]
> There is an alternative implementation of `sockstat` that ships with FreeBSD, but it might depend on parts of the BSD kernel which are not shared by XNU (the MacOS kernel).

If somebody still interested, try my implementation: https://github.com/mezantrop/sockstat

**Comment by iaminci** [comment, marked as answer]
> Hi,
> 
> I think it would be great if we could somehow install **netstat** used in linux distributions via brew as mac's netstat it's almost useless.
> 
> Thank you.

Hi, I have created netstat same as linux for macOS so you can try here is the link for the repository https://github.com/iaminci/homebrew-netstat-macos and suggest me if anything I can add in it.

**Reply by EricoCartmanez** [reply]
Good job!

**Reply by mezantrop** [reply]
So, this is a wrapper around lsof :)

**Reply by iaminci** [reply]
Yes, it is.
